src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,1840 @@
+
+  exprt char1 = str[minus_exprt(args[1],refined_string_typet::index_of_int(2))];
+  exprt char1_as_int = typecast_exprt(char1,return_type);
+  exprt char2 = str[minus_exprt(args[1],refined_string_typet::index_of_int(1))];
+  exprt char2_as_int = typecast_exprt(char2,return_type);
+
+  assert(return_type.id() == ID_signedbv);
+  size_t width = to_bitvector_type(return_type).get_width();
+
+  exprt pair_value =
+    plus_exprt
+    (constant_signed(0x010000,width),
+     (plus_exprt
+      (mult_exprt
+       (mod_exprt(char1_as_int,constant_signed(0x0800,width)),
+	constant_signed(0x0400,width)),
+       mod_exprt(char2_as_int,constant_signed(0x0400,width)))));
+
+  exprt return_pair = and_exprt(is_high_surrogate(char1),is_low_surrogate(char2));
+
+  axioms.emplace_back(return_pair,equal_exprt(result,pair_value));
+  axioms.emplace_back(not_exprt(return_pair),
+			     equal_exprt(result,char2_as_int));
+  return result;
+}
+
+exprt string_constraint_generatort::string_code_point_count
+( const function_application_exprt &f)
+{
+  string_exprt str = string_of_expr(args(f,3)[0]);
+  exprt begin = args(f,3)[1];
+  exprt end = args(f,3)[2];
+  typet return_type = f.type();
+  symbol_exprt result = string_exprt::fresh_symbol("code_point_count",return_type);
+  exprt length = minus_exprt(end,begin);
+  axioms.emplace_back(binary_relation_exprt(result,ID_le,length));
+  axioms.emplace_back(binary_relation_exprt(result,ID_ge,div_exprt(length,refined_string_typet::index_of_int(2))));
+
+  return result;
+}
+
+exprt string_constraint_generatort::string_offset_by_code_point
+( const function_application_exprt &f)
+{
+  string_exprt str = string_of_expr(args(f,3)[0]);
+  exprt index = args(f,3)[1];
+  exprt offset = args(f,3)[2];
+  typet return_type = f.type();
+  symbol_exprt result = string_exprt::fresh_symbol("offset_by_code_point",return_type);
+  exprt minimum = plus_exprt(index,plus_exprt(index,offset));
+  exprt maximum = plus_exprt(index,plus_exprt(index,mult_exprt(offset,refined_string_typet::index_of_int(2))));
+  axioms.emplace_back(binary_relation_exprt(result,ID_le,maximum));
+  axioms.emplace_back(binary_relation_exprt(result,ID_ge,minimum));
+
+  return result;
+}
+
+
+exprt string_constraint_generatort::string_to_char_array
+(const function_application_exprt &f)
+{
+  string_exprt str = string_of_expr(args(f,1)[0]);
+  return str.content();
+}
+
+
+exprt string_constraint_generatort::string_compare_to
+(const function_application_exprt &f)
+{
+  string_exprt s1 = string_of_expr(args(f,2)[0]);
+  string_exprt s2 = string_of_expr(args(f,2)[1]);
+  typet return_type = f.type();
+  symbol_exprt res = string_exprt::fresh_symbol("compare_to",return_type);
+
+  // In the lexicographic comparison, x is the first point where the two strings differ.
+  // res == 0 => |s1| = |s2| && forall i < |s1|. s1[i] == s2[i]
+  // res != 0 =>
+  //   (|s1| <= |s2| && exists x < |s1|. res = s1[x] - s2[x] && forall i<x s
```

Comment:
I haven't read all of these since they're quite stereotyped, but the same fixes apply as elsewhere -- bring the coding style in line with the codebase standard, and name the functions informatively so that at callsites it's clear that they're creating something rather than just fetching some existing thing.